docs/proposals/read-write-operations-bucket.md

```diff
@@ -0,0 +1,248 @@
+1. To schedule delete operation, delete `meta.json` file. All components will exclude this block and compactor will do eventual deletion assuming the block is partially uploaded. [Compactor change needed]
+1. Compactor waits minimum 15m (`deleteDelay`) before deleting the whole `To Delete` block. [Compactor change needed]
+
+Below diagram shows the block states defined by our rules.
+
+[![Diagram](https://docs.google.com/drawings/d/e/2PACX-1vSZjddrWBkf2ynczeIMSzGo-yqe7Cg9X-TG_LjyWAP_ToyZolMEG_UnbNlvoSetRhDLmlfABd_rJ4W9/pub?w=960&h=720)](https://docs.google.com/drawings/d/1leRN6GUvU98Ur2TowMefk4ycTFNBRi9t1-dGtOPmHg0/edit?usp=sharing)
+
+### Why?
+
+To deal with A and B, so overall eventual consistency, we will leverage existing feature of metric format (TSDB) which is 
+*block immutability*. In object storage format, a block is a directory named with [`ULID`](https://github.com/oklog/ulid) that contains
+`chunks` dir with one or more chunk files (each not bigger than 512MB), `index` and `meta.json` files. To ensure every component
+has the same state of metrics we create first strict rule:
+
+> 1 . A block, once uploaded is never edited (no file within ULID directory is removed, edited or added).
+
+With this rule alone we restrict bucket operations only to:
+* Block upload
+* Block removal 
+
+As those operations are not atomic itself, partial blocks or inconsistent block view may occur for short time (read-after-write consistency).
+To mitigate this we introduce second rule:
+
+ > 2 . A block is assumed ready to be read *only* after 15 minutes (configured by `syncDelay`) from *creation time*. Block id (`ULID`) is used
+to determine creation time (`if ulid.Now()-id.Time() < uint64(c.syncDelay/time.Millisecond)`). Block fresher than 15 minutes is assumed
+still in upload (`Fresh`). This can be overridden by `ignore_delay` field in meta.json.
+
+Readers may attempt to load block's files (e.g meta.json to grab external labels), but should be aware that those files can be partially uploaded.
+
+The delay is not only for block eventual consistency. Even if everyone can read block properly, we are not aware if all readers loaded the block properly.
+
+To ensure we have block fully uploaded we define:
+
+> 3 . TSDB `meta.json` file within block *have to be* uploaded as first one.
+
+Currently and as proposed `meta.json` is our integrity marker (it is uploaded as the last one). It matters even when we don't have strong consistency, because 
+it helps us detect the case of `upload` crashed in between of upload. We only upload meta.json after successful uploads of rest of the files for the block.
+
+With those rule one could say "why to wait 15 minutes if we can just check if meta.json is parsable". In eventual consistency world
+we cannot just check meta.json as even when it was uploaded at the end, other files might be still unavailable for a reader.
+
+Thanks of meta.json being uploaded at the end we can form 4th rule:
```

Comment:
Maybe I just don't understand it, but in:
* in l.93 you stated meta.json is uploaded first
* in l.101 you stated meta.json being uploaded at the end

Reply:
Should indeed be the last thing was a typo earlier in the doc, think it will be corrected soon :+1:

Reply:
Yea, it is a typo sorry ;p Meta.json should be uploaded as last